File: src/lineage/simulation.py
"""Simulation engine: tick loop, perception, actions, energy economy."""
from __future__ import annotations

import math
import random
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from .genome import BehaviorAction, BehaviorNodeType, Genome, crossover, mutate
from .speciation import SpeciesTracker
from .world import Food, Organism, Position, World, WorldConfig

if TYPE_CHECKING:
    from .config import SimulationConfig


@dataclass
class SimulationEvent:
    tick: int
    message: str
    event_type: str = "info"


class Simulation:
    """Main simulation loop."""

    def __init__(self, config: SimulationConfig) -> None:
        self.config = config
        self.world = World(config=config.world)
        self.species_tracker = SpeciesTracker(
            threshold=config.world.speciation_threshold
        )
        self.events: list[SimulationEvent] = []
        self.rng = random.Random(config.seed)
        self._tick = 0

    def initialize(self) -> None:
        self.world.initialize()
        for org in self.world.organisms:
            self.species_tracker.assign_species(org, tick=0)

    def run_tick(self) -> dict[str, float | int]:
        self._tick += 1
        self.world.tick_count = self._tick

        self.world.rebuild_spatial_hash()

        if self.rng.random() < self.world.config.food_spawn_rate:
            self.world.spawn_food()

        for org in list(self.world.organisms):
            if not org.alive:
                continue

            self._process_organism(org)

        self.world.remove_dead()

        speciation_events = self.species_tracker.update_species(
            self.world.organisms, self._tick
        )
        for msg in speciation_events:
            self.events.append(
                SimulationEvent(self._tick, msg, "speciation")
            )

        for f in self.world.food:
            f.age += 1

        self._apply_carrying_capacity()

        return self.world.statistics()

    def _process_organism(self, org: Organism) -> None:
        """Process a single organism for one tick."""
        org.tick()

        org.energy -= org.metabolism_cost()
        org.energy += org.photosynthesis_gain()

        if org.energy <= 0:
            org.mark_dead()
            return

        perception = self._perceive(org)
        action = self._decide(org, perception)
        self._act(org, action, perception)

        org.energy = min(org.energy, 200.0)

    def _perceive(self, org: Organism) -> dict[str, Any]:
        """Gather sensory information for an organism."""
        pos = org.position
        width, height = self.world.width(), self.world.height()

        nearby_food = self.world.nearby_food(
            pos, org.genome.food_detection_radius
        )
        nearby_organisms = self.world.nearby_organisms(
            pos, max(
                org.genome.predator_detection_radius,
                org.genome.mate_detection_radius,
            ),
            exclude_id=org.id,
        )

        predators = [
            o for o in nearby_organisms
            if o.genome.size > org.genome.size * 1.2
        ]
        mates = [
            o for o in nearby_organisms
            if o.genome.can_interbreed(
                org.genome, self.world.config.speciation_threshold
            )
        ]

        closest_food = None
        closest_food_dist = float("inf")
        for f in nearby_food:
            d = pos.distance_to(f.position, width, height)
            if d < closest_food_dist:
                closest_food_dist = d
                closest_food = f

        closest_predator = None
        closest_predator_dist = float("inf")
        for p in predators:
            d = pos.distance_to(p.position, width, height)
            if d < closest_predator_dist:
                closest_predator_dist = d
                closest_predator = p

        closest_mate = None
        closest_mate_dist = float("inf")
        for m in mates:
            d = pos.distance_to(m.position, width, height)
            if d < closest_mate_dist:
                closest_mate_dist = d
                closest_mate = m

        return {
            "food_nearby": closest_food is not None,
            "food": closest_food,
            "food_distance": closest_food_dist,
            "predator_nearby": closest_predator is not None,
            "predator": closest_predator,
            "predator_distance": closest_predator_dist,
            "mate_nearby": closest_mate is not None,
            "mate": closest_mate,
            "mate_distance": closest_mate_dist,
            "energy": org.energy,
            "position": pos,
        }

    def _decide(
        self, org: Organism, perception: dict[str, Any]
    ) -> BehaviorAction:
        """Evaluate behavior tree to decide action."""
        tree = org.genome.behavior_tree
        if not tree:
            return BehaviorAction.WANDER

        node_idx = 0
        while True:
            if node_idx >= len(tree):
                return BehaviorAction.WANDER

            node = tree[node_idx]
            condition_met = self._evaluate_condition(
                node.condition, node.threshold, perception
            )

            if condition_met:
                return node.if_true
            else:
                if node.is_leaf:
                    if isinstance(node.if_false, BehaviorAction):
                        return node.if_false
                    return BehaviorAction.WANDER
                else:
                    if isinstance(node.if_false, int):
                        node_idx = node.if_false
                    else:
                        return node.if_false

    def _evaluate_condition(
        self,
        condition: BehaviorNodeType,
        threshold: float,
        perception: dict[str, Any],
    ) -> bool:
        if condition == BehaviorNodeType.ALWAYS:
            return True
        if condition == BehaviorNodeType.FOOD_NEARBY:
            return perception["food_nearby"] and perception["food_distance"] <= threshold + 5
        if condition == BehaviorNodeType.PREDATOR_NEARBY:
            return perception["predator_nearby"] and perception["predator_distance"] <= threshold + 5
        if condition == BehaviorNodeType.MATE_NEARBY:
            return perception["mate_nearby"] and perception["mate_distance"] <= threshold + 5
        if condition == BehaviorNodeType.ENERGY_ABOVE:
            return perception["energy"] > threshold
        return False

    def _act(
        self,
        org: Organism,
        action: BehaviorAction,
        perception: dict[str, Any],
    ) -> None:
        width, height = self.world.width(), self.world.height()

        if action == BehaviorAction.REST:
            pass

        elif action == BehaviorAction.WANDER:
            angle = self.rng.uniform(0, 2 * math.pi)
            speed = org.speed()
            dx = math.cos(angle) * speed
            dy = math.sin(angle) * speed
            org.position.move(dx, dy, width, height)
            org.energy -= speed * 0.05

        elif action == BehaviorAction.MOVE_TOWARD_FOOD:
            food = perception.get("food")
            if food:
                dx, dy = org.position.direction_to(
                    food.position, width, height
                )
                speed = org.speed()
                org.position.move(dx * speed, dy * speed, width, height)
                org.energy -= speed * 0.05
                dist = org.position.distance_to(food.position, width, height)
                if dist < 1.0 and food in self.world.food:
                    org.energy += food.energy
                    self.world.food.remove(food)
            else:
                angle = self.rng.uniform(0, 2 * math.pi)
                speed = org.speed()
                org.position.move(
                    math.cos(angle) * speed,
                    math.sin(angle) * speed,
                    width,
                    height,
                )
                org.energy -= speed * 0.05

        elif action == BehaviorAction.MOVE_TOWARD_MATE:
            mate = perception.get("mate")
            if mate and mate.alive:
                dx, dy = org.position.direction_to(
                    mate.position, width, height
                )
                speed = org.speed()
                org.position.move(dx * speed, dy * speed, width, height)
                org.energy -= speed * 0.05
                dist = org.position.distance_to(mate.position, width, height)
                if dist < 1.0:
                    self._attempt_reproduction(org, mate)
            else:
                angle = self.rng.uniform(0, 2 * math.pi)
                speed = org.speed()
                org.position.move(
                    math.cos(angle) * speed,
                    math.sin(angle) * speed,
                    width,
                    height,
                )
                org.energy -= speed * 0.05

        elif action == BehaviorAction.FLEE:
            predator = perception.get("predator")
            if predator:
                dx, dy = org.position.direction_to(
                    predator.position, width, height
                )
                speed = org.speed() * 1.5
                org.position.move(-dx * speed, -dy * speed, width, height)
                org.energy -= speed * 0.08
            else:
                angle = self.rng.uniform(0, 2 * math.pi)
                speed = org.speed()
                org.position.move(
                    math.cos(angle) * speed,
                    math.sin(angle) * speed,
                    width,
                    height,
                )
                org.energy -= speed * 0.05

        elif action == BehaviorAction.REPRODUCE:
            if org.can_reproduce():
                self._asexual_reproduction(org)
            else:
                angle = self.rng.uniform(0, 2 * math.pi)
                speed = org.speed()
                org.position.move(
                    math.cos(angle) * speed,
                    math.sin(angle) * speed,
                    width,
                    height,
                )
                org.energy -= speed * 0.05

        elif action == BehaviorAction.MOVE_TOWARD_CENTER:
            cx, cy = width / 2, height / 2
            dx, dy = org.position.direction_to(
                Position(cx, cy), width, height
            )
            speed = org.speed()
            org.position.move(dx * speed, dy * speed, width, height)
            org.energy -= speed * 0.05

    def _attempt_reproduction(
        self, parent_a: Organism, parent_b: Organism
    ) -> None:
        """Attempt sexual reproduction between two organisms."""
        if not parent_a.can_reproduce() or not parent_b.can_reproduce():
            return
        if not parent_a.genome.can_interbreed(
            parent_b.genome, self.world.config.speciation_threshold
        ):
            return

        child_genome = crossover(parent_a.genome, parent_b.genome)
        child_genome = mutate(child_genome)

        parent_a.energy -= parent_a.reproduce_cost()
        parent_b.energy -= parent_b.reproduce_cost()
        parent_a.ticks_since_reproduction = 0
        parent_b.ticks_since_reproduction = 0

        child = self.world.spawn_organism(
            genome=child_genome,
            position=Position(parent_a.position.x, parent_a.position.y),
        )
        self.species_tracker.assign_species(child, self._tick)

    def _asexual_reproduction(self, parent: Organism) -> None:
        """Asexual reproduction: clone + mutate."""
        child_genome = mutate(parent.genome)
        parent.energy -= parent.reproduce_cost()
        parent.ticks_since_reproduction = 0

        child = self.world.spawn_organism(
            genome=child_genome,
            position=Position(parent.position.x, parent.position.y),
        )
        self.species_tracker.assign_species(child, self._tick)

    def _apply_carrying_capacity(self) -> None:
        current_pop = len([o for o in self.world.organisms if o.alive])
        max_pop = self.world.config.max_population
        ratio = current_pop / max_pop
        if ratio > 0.7:
            pressure = (ratio - 0.7) / 0.3
            for org in self.world.organisms:
                if org.alive:
                    org.energy -= org.metabolism_cost() * pressure * 2.0
                    if self.rng.random() < pressure * 0.3:
                        org.energy -= 2.0
                    if org.energy <= 0:
                        org.mark_dead()

    def run(self, max_ticks: int | None = None) -> None:
        """Run the simulation for the specified number of ticks."""
        if max_ticks is None:
            max_ticks = self.config.ticks

        for _ in range(max_ticks):
            if self._tick >= max_ticks:
                break
            self.run_tick()

    def get_events(self) -> list[SimulationEvent]:
        return self.events

    def to_dict(self) -> dict:
        return {
            "tick": self._tick,
            "world": self.world.to_dict(),
            "species": self.species_tracker.to_dict(),
            "events": [
                {"tick": e.tick, "message": e.message, "type": e.event_type}
                for e in self.events[-100:]
            ],
        }

File: web/src/simulation/simulation.ts
/**
 * Simulation engine: tick loop, perception, actions, energy economy.
 * Port of Python src/lineage/simulation.py — matches behavior exactly.
 *
 * Tick processing order (CRITICAL — must not change):
 * 1. Increment tick count
 * 2. Rebuild spatial hash
 * 3. Spawn food (random chance based on food_spawn_rate)
 * 4. Process each organism (metabolism, photosynthesis, perception, decision, action)
 * 5. Remove dead organisms
 * 6. Update species (speciation / extinction events)
 * 7. Age food
 * 8. Apply carrying capacity pressure
 * 9. Return statistics
 */

import { World, Position } from "./world";
import { PRNG } from "../utils/prng";
import { crossover, mutate, canInterbreed } from "./genome";
import {
  decideAction,
  act,
  type Perception,
  type WorldContext,
} from "./behavior";
import type {
  SimulationConfig,
  WorldStatistics,
  SimulationEvent,
  Organism as OrganismType,
  Food as FoodType,
} from "./types";
import { SpeciesTracker } from "./speciation";

const ENERGY_CAP = 200.0;

type WorldOrganism = ReturnType<World["spawnOrganism"]>;

export class Simulation {
  config: SimulationConfig;
  world: World;
  speciesTracker: SpeciesTracker;
  events: SimulationEvent[];
  rng: PRNG;
  private _tick: number;

  constructor(config: SimulationConfig) {
    this.config = config;
    this.world = new World(config);
    this.speciesTracker = new SpeciesTracker(config.speciationThreshold);
    this.events = [];
    this.rng = new PRNG(config.seed);
    this._tick = 0;
  }

  initialize(): void {
    this.world.initialize();
    for (const org of this.world.organisms) {
      this.speciesTracker.assignSpecies(org, 0);
    }
  }

  runTick(): WorldStatistics {
    this._tick += 1;
    this.world.tickCount = this._tick;

    this.world.rebuildSpatialHash();

    if (this.rng.random() < this.config.foodSpawnRate) {
      this.world.spawnFood();
    }

    const snapshot = [...this.world.organisms];
    for (const org of snapshot) {
      if (!org.alive) continue;
      this.processOrganism(org);
    }

    this.world.removeDead();

    const speciationMessages = this.speciesTracker.updateSpecies(
      this.world.organisms,
      this._tick,
    );
    for (const msg of speciationMessages) {
      const eventType = msg.toLowerCase().includes("extinct")
        ? "extinction"
        : "speciation";
      this.events.push({ tick: this._tick, message: msg, eventType });
    }

    for (const f of this.world.food) {
      f.age += 1;
    }

    this.applyCarryingCapacity();

    return this.world.statistics();
  }

  private processOrganism(org: WorldOrganism): void {
    org.tick();
    org.energy -= org.metabolismCost();
    org.energy += org.photosynthesisGain();

    if (org.energy <= 0) {
      org.markDead();
      return;
    }

    const perception = this.perceive(org);
    const action = decideAction(org as unknown as OrganismType, perception);

    const worldContext: WorldContext = {
      width: this.config.worldWidth,
      height: this.config.worldHeight,
      food: this.world.food as unknown as FoodType[],
      organisms: this.world.organisms as unknown as OrganismType[],
      speciationThreshold: this.config.speciationThreshold,
    };

    act(
      org as unknown as OrganismType,
      action,
      perception,
      worldContext,
      this.rng,
      (parentA, parentB) => {
        const a = parentA as unknown as WorldOrganism;
        if (parentB !== undefined) {
          this.attemptReproduction(a, parentB as unknown as WorldOrganism);
        } else {
          this.asexualReproduction(a);
        }
      },
    );

    org.energy = Math.min(org.energy, ENERGY_CAP);
  }

  private perceive(org: WorldOrganism): Perception {
    const pos = org.position;
    const width = this.config.worldWidth;
    const height = this.config.worldHeight;

    const nearbyFoodResult = this.world.nearbyFood(
      pos,
      org.genome.foodDetectionRadius,
    );
    const detectionRadius = Math.max(
      org.genome.predatorDetectionRadius,
      org.genome.mateDetectionRadius,
    );
    const nearbyOrganismsResult = this.world.nearbyOrganisms(
      pos,
      detectionRadius,
      org.id,
    );

    const predators = nearbyOrganismsResult.filter(
      (o) => o.genome.size > org.genome.size * 1.2,
    );
    const preyList = nearbyOrganismsResult.filter(
      (o) => o.genome.size < org.genome.size * 0.8 && o.id !== org.id,
    );
    const mates = nearbyOrganismsResult.filter((o) =>
      canInterbreed(o.genome, org.genome, this.config.speciationThreshold),
    );

    let closestFood: (typeof nearbyFoodResult)[number] | null = null;
    let closestFoodDist = Infinity;
    for (const f of nearbyFoodResult) {
      const d = pos.distanceTo(f.position, width, height);
      if (d < closestFoodDist) {
        closestFoodDist = d;
        closestFood = f;
      }
    }

    let closestPredator: (typeof nearbyOrganismsResult)[number] | null = null;
    let closestPredatorDist = Infinity;
    for (const p of predators) {
      const d = pos.distanceTo(p.position, width, height);
      if (d < closestPredatorDist) {
        closestPredatorDist = d;
        closestPredator = p;
      }
    }

    let closestPrey: (typeof nearbyOrganismsResult)[number] | null = null;
    let closestPreyDist = Infinity;
    for (const pr of preyList) {
      const d = pos.distanceTo(pr.position, width, height);
      if (d < closestPreyDist) {
        closestPreyDist = d;
        closestPrey = pr;
      }
    }

    let closestMate: (typeof nearbyOrganismsResult)[number] | null = null;
    let closestMateDist = Infinity;
    for (const m of mates) {
      const d = pos.distanceTo(m.position, width, height);
      if (d < closestMateDist) {
        closestMateDist = d;
        closestMate = m;
      }
    }

    return {
      foodNearby: closestFood !== null,
      food: closestFood as unknown as FoodType | null,
      foodDistance: closestFoodDist,
      predatorNearby: closestPredator !== null,
      predator: closestPredator as unknown as OrganismType | null,
      predatorDistance: closestPredatorDist,
      preyNearby: closestPrey !== null,
      prey: closestPrey as unknown as OrganismType | null,
      preyDistance: closestPreyDist,
      mateNearby: closestMate !== null,
      mate: closestMate as unknown as OrganismType | null,
      mateDistance: closestMateDist,
      energy: org.energy,
      position: { x: pos.x, y: pos.y },
    };
  }

  private attemptReproduction(
    parentA: WorldOrganism,
    parentB: WorldOrganism,
  ): void {
    if (!parentA.canReproduce() || !parentB.canReproduce()) return;
    if (
      !canInterbreed(
        parentA.genome,
        parentB.genome,
        this.config.speciationThreshold,
      )
    ) {
      return;
    }

    const childGenome = mutate(
      crossover(parentA.genome, parentB.genome, this.rng),
      this.rng,
    );

    parentA.energy -= parentA.reproduceCost();
    parentB.energy -= parentB.reproduceCost();
    parentA.ticksSinceReproduction = 0;
    parentB.ticksSinceReproduction = 0;

    const child = this.world.spawnOrganism(
      childGenome,
      new Position(parentA.position.x, parentA.position.y),
    );

    const [newSid, isNew] = this.speciesTracker.assignSpecies(
      child,
      this._tick,
    );
    if (isNew) {
      this.events.push({
        tick: this._tick,
        message: `New species ${newSid} born from parents ${parentA.genome.speciesId} and ${parentB.genome.speciesId}`,
        eventType: "speciation",
      });
    }
  }

  private asexualReproduction(parent: WorldOrganism): void {
    const childGenome = mutate(parent.genome, this.rng);

    parent.energy -= parent.reproduceCost();
    parent.ticksSinceReproduction = 0;

    const child = this.world.spawnOrganism(
      childGenome,
      new Position(parent.position.x, parent.position.y),
    );

    const [newSid, isNew] = this.speciesTracker.assignSpecies(
      child,
      this._tick,
    );
    if (isNew) {
      this.events.push({
        tick: this._tick,
        message: `New species ${newSid} diverged from species ${parent.genome.speciesId} (asexual)`,
        eventType: "speciation",
      });
    }
  }

  private applyCarryingCapacity(): void {
    const aliveCount = this.world.organisms.filter((o) => o.alive).length;
    const maxPop = this.config.maxPopulation;
    const ratio = aliveCount / maxPop;

    if (ratio > 0.7) {
      const pressure = (ratio - 0.7) / 0.3;
      for (const org of this.world.organisms) {
        if (!org.alive) continue;
        org.energy -= org.metabolismCost() * pressure * 2.0;
        if (this.rng.random() < pressure * 0.3) {
          org.energy -= 2.0;
        }
        if (org.energy <= 0) {
          org.markDead();
        }
      }
    }
  }

  run(maxTicks?: number): void {
    const limit = maxTicks ?? this.config.ticks;
    for (let i = 0; i < limit; i++) {
      if (this._tick >= limit) break;
      this.runTick();
    }
  }

  getEvents(): SimulationEvent[] {
    return this.events;
  }

  getState(): { tick: number; organisms: Array<{ id: number; x: number; y: number; energy: number; size: number; hue: number; shape: string; speciesId: number; alive: boolean }>; food: Array<{ x: number; y: number; energy: number }>; speciesCount: number } {
    return {
      tick: this._tick,
      organisms: this.world.organisms.map((o) => ({
        id: o.id,
        x: o.position.x,
        y: o.position.y,
        energy: o.energy,
        size: o.genome.size,
        hue: o.genome.colorHue,
        shape: o.genome.shape,
        speciesId: o.genome.speciesId,
        alive: o.alive,
      })),
      food: this.world.food.map((f) => ({
        x: f.position.x,
        y: f.position.y,
        energy: f.energy,
      })),
      speciesCount: this.speciesTracker.getSpeciesCount(),
    };
  }

  toDict(): Record<string, unknown> {
    return {
      tick: this._tick,
      world: this.world.toDict(),
      species: this.speciesTracker.toDict(),
      events: this.events.slice(-100).map((e) => ({
        tick: e.tick,
        message: e.message,
        type: e.eventType,
      })),
    };
  }
}
